homeassistant/components/roborock/__init__.py

```diff
@@ -40,61 +42,103 @@ async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
-            network_info[device_id],
-            product_info[device.product_id],
-            mqtt_clients[device.duid],
-        )
-    await asyncio.gather(
-        *(coordinator.verify_api() for coordinator in coordinator_map.values())
-    )
-    # If one device update fails - we still want to set up other devices
-    await asyncio.gather(
-        *(
-            coordinator.async_config_entry_first_refresh()
-            for coordinator in coordinator_map.values()
-        ),
+    # Get a Coordinator if the device is available or if we have connected to the device before
+    coordinators = await asyncio.gather(
+        *build_setup_functions(hass, device_map, user_data, product_info),
         return_exceptions=True,
     )
+    # Valid coordinators are those where we had networking cached or we could get networking
+    valid_coordinators: list[RoborockDataUpdateCoordinator] = [
+        coord
+        for coord in coordinators
+        if isinstance(coord, RoborockDataUpdateCoordinator)
+    ]
+    if len(valid_coordinators) == 0:
+        raise ConfigEntryNotReady("No coordinators were able to successfully setup.")
     hass.data.setdefault(DOMAIN, {})[entry.entry_id] = {
-        device_id: coordinator
-        for device_id, coordinator in coordinator_map.items()
-        if coordinator.last_update_success
-    }  # Only add coordinators that succeeded
-
-    if not hass.data[DOMAIN][entry.entry_id]:
-        # Don't start if no coordinators succeeded.
-        raise ConfigEntryNotReady("There are no devices that can currently be reached.")
-
+        coordinator.roborock_device_info.device.duid: coordinator
+        for coordinator in valid_coordinators
+    }
     await hass.config_entries.async_forward_entry_setups(entry, PLATFORMS)
 
     return True
 
 
+def build_setup_functions(
+    hass: HomeAssistant,
+    device_map: dict[str, HomeDataDevice],
+    user_data: UserData,
+    product_info: dict[str, HomeDataProduct],
+) -> list[Coroutine[Any, Any, RoborockDataUpdateCoordinator | None]]:
+    """Create a list of setup functions that can later be called asynchronously."""
+    setup_functions = []
+    for device in device_map.values():
+        setup_functions.append(
+            setup_device(hass, user_data, device, product_info[device.product_id])
+        )
+    return setup_functions
+
+
+async def setup_device(
+    hass: HomeAssistant,
+    user_data: UserData,
+    device: HomeDataDevice,
+    product_info: HomeDataProduct,
+) -> RoborockDataUpdateCoordinator | None:
+    """Set up a device Coordinator."""
+    mqtt_client = RoborockMqttClient(user_data, DeviceData(device, product_info.name))
+    try:
+        networking = await mqtt_client.get_networking()
+        if networking is None:
+            # If the api does not return an error but does return None for
+            # get_networking - then we need to go through cache checking.
+            raise RoborockException("Networking request returned None.")
```

Comment:
We should not raise an exception from an external library, unless it's a helper library that uses exceptions as part of its interface, like voluptuous. Home Assistant should normally only raise exceptions inheriting `HomeAssistantError`.